Fix interpolation with repeating, mutating names

mortenpi opened the pull request:
A variable that is interpolated multiple times with updates in between will print with the final value in all positions:

```
julia> let x = 1
           function f!(); x += 1; 42; end
           cm"$(x), $(f!()), $(x)"
       end
 2, 42, 2
```

This is because `JuliaValue` is a struct and therefore the two instance of `JuliaValue(:x, nothing)` are identical. This confuses the replacement `Dict` in `_interp!`, where value from the earlier position gets overwritten.

Changing `JuliaValue` to be mutable and using `IdDict` in `_interp!` fixes this:

```
julia> let x = 1
           function f!(); x += 1; 42; end
           cm"$(x), $(f!()), $(x)"
       end
 1, 42, 2
```

A few notes:

* Strictly, `IdDict` is not necessary. However, using it makes sure it keeps working if `==` and `hash` were to become overloaded (which is what I do in MarkdownAST right now).

* An alternative solution could be to add a position counter to `JuliaValue`, which would get incremented every time the parser constructs a new `JuliaValue`.

* This only affects pure variable interpolation, since `Expr` objects are mutable, and so won't match even if they have identical content.

MichaelHatherly commented:
Nice catch.

> An alternative solution could be to add a position counter to JuliaValue, which would get incremented every time the parser constructs a new JuliaValue.

I think I'm more partial to a counter-based approach to this rather than making `JuliaValue` mutable, but if that implementation would be much more complex than the current one then we can just go with this `IdDict` approach.

mortenpi commented:
I've updated this with a counter-based approach. This actually allows us to get rid of the `Dict` altogether, since we're literally storing the array indices in the counter now.

However, rather than just adding the `.pos` field to `JuliaValue`, I decided to create a new `JuliaExpression` inline element that gets added to the tree during parsing, and then replaced by `JuliaValue`. This has two benefits in my mind:

1. It allows us to discard the position information, since that becomes irrelevant once the values have been evaluated.
2. It makes it explicit whether it's an expression that has been evaluated or not (in case someone runs the interpolation parser without the `cm""` macro). Because of this, I also removed the one-argument constructor for `JuliaValue`, implying that a `JuliaValue` must always be an evaluated value.

MichaelHatherly commented on src/extensions/interpolation.jl:
```diff
@@ -4,19 +4,24 @@
+    pos::Int
+    ex
+    JuliaExpression(pos, ex) = new(pos, ex)
+end
```
Should there be the `write_*` methods defined on `JuliaExpression` since it is an `AbstractInline`? If any values of those types were to leak into a final AST (if they can even?) what gets printed at the moment?

mortenpi replied:
Ah, this is a fair point, I forgot about he `write_*` methods. In principle, as long as you use the `cm""` macro, I don't think it should be possible for `JuliaExpression` to leak, but they would end up in the AST if you construct your own parser with a `JuliaInterpolationRule`.

I think it would be reasonable to have the methods defined, and I guess they should do what the Markdown writer does right now and just print the expression with `$(...)` around it?

MichaelHatherly replied:
> I guess they should do what the Markdown writer does right now and just print the expression with `$(...)` around it?

Sounds reasonable.

MichaelHatherly reviewed (approved):
This seems like the right approach to take.

mortenpi commented:
Added the writer methods.

mortenpi commented on src/extensions/interpolation.jl:
```diff
@@ -164,6 +167,27 @@ end
+    tag(rend, "/span")
+end
+
+function write_latex(jv::JuliaExpression, rend, node, enter)
+    print(rend.buffer, sprint(print, '\\', '$', "($(jv.ex))"))
+end
```
Actually I just realized the LaTeX is problematic, since it would get interpreted as valid LaTeX commands..

MichaelHatherly replied:
Ah, maybe just escape the printed expression. Pretty sure there's some latex escaping code somewhere in the package...

mortenpi replied:
Fixed it to use `latex_escape`, as is done for `Code`.

MichaelHatherly commented:
Thanks Morten!
